src/dotnet/CarbonAware/IntervalHelper.cs

```diff
@@ -0,0 +1,66 @@
+namespace CarbonAware;
+
+/// <summary>
+/// Helper for intervals within Carbon Aware.
+/// </summary>
+public static class IntervalHelper
+{
+
+    /// <summary>
+    /// Ensures the data is valid by using the min sampling window. Uses the expanded data to add an extra
+    /// data point before the startDate to ensure all expected data is included.
+    /// </summary>
+    /// <param name="data">Expanded data that includes the minimum sampling window</param>
+    /// <param name="startDate">Original start date provided by user</param>
+    /// <param name="endDate">Original end date provided by user</param>
+    /// <returns>Filtered emissions data.</returns>
+    /// <remarks> If the user provided interval is already >= minimum sampling window, or no expanded data, we just return the passed data. </remarks>
+    public static IEnumerable<EmissionsData> MinSamplingFiltering(IEnumerable<EmissionsData> expandedData, DateTimeOffset startDate, DateTimeOffset endDate, double minSamplingWindow)
+    {
+        TimeSpan ts = endDate - startDate;
+        if ((ts.TotalMinutes >= minSamplingWindow) || !expandedData.Any())
+        {
+            return expandedData;
+        }
+
+        var arrData = expandedData.ToArray();
+        var startDateTime = startDate.DateTime;
+        var dataLength = arrData.Length;
+        var splitIndex = dataLength - 1;
+
+        // sort data since different sources might have populated the data differently.
+        Array.Sort(arrData, new CompareEmissionDataSort());
+
+        // Find index of data point right before startDateTime
+        while (!(arrData[splitIndex].Time < startDateTime)){
+            if (splitIndex == 0) break;
+            splitIndex --;
+        }
+
+        // copy from the split index on
+        var filteredData = new EmissionsData[dataLength - splitIndex];
+        Array.Copy(arrData, splitIndex, filteredData, 0, filteredData.Length);
```

Comment:
This logic works only to filter data before the `startTime`,  it will still return data beyond the requested `endTime`.

Reply:
Right now we are only moving the start time, since that is the one problematic , when the range doesn't include content. Moving both don't make sense, since the idea is to capture at least one item closer to start time, which would be the one that has more 'relevant' points.

Reply:
Hey @bderusha , after some more thoughts with @JenMadiedo , seems like a better approach would be to always expand the window from both ends (start and end times) and filter till the closest initial times. That way it would cover both ends. Let us know your thoughts.